### PR TITLE
DPR-311: Apply CDC events to structured zone

### PR DESCRIPTION
**Pull request by koladeadewuyi-moj**
This PR:
- Applies the incremental/CDC events to the structured zone
- The CDC events are applied (inserted/updated/deleted) on a row-by-row basis
- In case a row operation fails, the row data and destination table are logged without failing the spark job

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/converter/dms/DMS_3_4_6.java [hunk 1]
```diff
@@ -78,12 +77,7 @@ public static Optional<Operation> getOperation(String operation) {
             return Arrays.stream(values()).filter(it -> it.name().equalsIgnoreCase(operation)).findAny();
         }
 
-        private static final DMS_3_4_6.Operation[] cdcOperations = { Insert, Update, Delete };
-
-        public static final Set<String> cdcOperationsSet = Arrays
-                .stream(cdcOperations)
-                .map(DMS_3_4_6.Operation::getName)
-                .collect(Collectors.toSet());
+        public static final Object[] cdcOperations = { Insert.getName(), Update.getName(), Delete.getName() };
```
Comment (kartthik-n) [inline src/main/java/uk/gov/justice/digital/converter/dms/DMS_3_4_6.java:80]: do we need Enum here? Final variables not enough?

File: src/main/java/uk/gov/justice/digital/job/DataHubJob.java [hunk 2]
```diff
@@ -94,8 +94,8 @@ private void batchProcessor(JavaRDD<byte[]> batch) {
                     val structuredLoadDataFrame = structuredZone.processLoad(spark, dataFrameForTable, table);
                     curatedZone.processLoad(spark, structuredLoadDataFrame, table);
 
-                    val structuredCDCDataFrame = structuredZone.processCDC(spark, dataFrameForTable, table);
-                    curatedZone.processCDC(spark, structuredCDCDataFrame, table);
+                    val structuredIncrementalDataFrame = structuredZone.processCDC(spark, dataFrameForTable, table);
```
Comment (kartthik-n) [inline src/main/java/uk/gov/justice/digital/job/DataHubJob.java:97]: there is not need for procesLoad and processCDC methods. we can call process() method directly passing the boolean flag to identify isCDC

File: src/main/java/uk/gov/justice/digital/zone/StructuredZone.java [hunk 3]
```diff
@@ -211,9 +230,49 @@ private void appendDataAndUpdateManifestForTable(SparkSession spark,
...
+                                storage.appendDistinct(destinationPath, dataFrame, primaryKey);
+                                break;
+                            case Update:
+                                storage.update(destinationPath, dataFrame, primaryKey);
+                                break;
+                            case Delete:
+                                storage.deleteRecords(destinationPath, dataFrame, primaryKey);
+                                break;
+                            default:
+                                break;
```
Comment (kartthik-n) [inline src/main/java/uk/gov/justice/digital/zone/StructuredZone.java:266]: default block doesn't need a exception here?

Comment (kartthik-n) [reply src/main/java/uk/gov/justice/digital/zone/StructuredZone.java:266]: also please add a log statement into default block or throw exception

File: src/test/java/uk/gov/justice/digital/zone/Fixtures.java [hunk 4]
```diff
@@ -19,35 +21,88 @@ public class Fixtures {
...
     public static final String TABLE_OPERATION = "load";
     public static final String ROW_CONVERTER = "row_converter";
-    public static final String RAW_DATA = "{}";
 
-    public static final StructType ROW_SCHEMA = new StructType()
-            .add(SOURCE, StringType, false)
-            .add(TABLE, StringType, false)
-            .add(OPERATION, StringType, false);
+    public static final String RECORD_KEY_1 = "record-1";
+    public static final String RECORD_KEY_2 = "record-2";
```
Comment (kartthik-n) [inline src/test/java/uk/gov/justice/digital/zone/Fixtures.java:26]: is that modifier can be private here?

File: src/test/java/uk/gov/justice/digital/zone/Fixtures.java [hunk 5]
```diff
@@ -19,35 +21,88 @@ public class Fixtures {
...
-        );
+    public static final String PRIMARY_KEY_PLACEHOLDER = "<PRIMARY-KEY>";
 
-    public static final GenericRowWithSchema dataMigrationEventRowWithInvalidOperation = new GenericRowWithSchema(
-            new Object[] { TABLE_SOURCE, TABLE_NAME, "makeTea" },
-            ROW_SCHEMA
-    );
+    public static String PRIMARY_KEY_FIELD = "primary-key";
 
+    public static String STRING_FIELD_KEY = "string-key";
```
Comment (kartthik-n) [inline src/test/java/uk/gov/justice/digital/zone/Fixtures.java:37]: can we move all these final values to abstract class and no need of private constructor Fixtures()?

File: src/main/java/uk/gov/justice/digital/job/DataHubJob.java (hunk 2 above)
Comment (koladeadewuyi-moj) [reply src/main/java/uk/gov/justice/digital/job/DataHubJob.java]: This is now updated. Thanks

File: src/main/java/uk/gov/justice/digital/zone/StructuredZone.java (hunk 3 above)
Comment (koladeadewuyi-moj) [reply src/main/java/uk/gov/justice/digital/zone/StructuredZone.java:263]: Added a log statement 👍

File: src/test/java/uk/gov/justice/digital/zone/Fixtures.java (hunk 5 above)
Comment (koladeadewuyi-moj) [reply src/test/java/uk/gov/justice/digital/zone/Fixtures.java]: Done 👍

File: src/test/java/uk/gov/justice/digital/zone/Fixtures.java (hunk 4 above)
Comment (koladeadewuyi-moj) [reply src/test/java/uk/gov/justice/digital/zone/Fixtures.java]: Updated. Thanks

**Review by kartthik-n — APPROVED** [review]

**Pull request merged**